pep-9999.rst

```diff
@@ -0,0 +1,252 @@
+considers the following terms and definitions:
+
+type variable
+    An instance of ``typing.TypeVar`` or ``typing.ParamSpec``, as defined
+    in `PEP 484 <https://www.python.org/dev/peps/pep-0484/>`_ and 
+    PEP `PEP 612 <https://www.python.org/dev/peps/pep-0612/>`_. Type 
+    variables can be used as either type annotations, or to declare 
+    user-defined generic types.
+
+type parameter
+    Generic types can have one or multiple type parameters. For user-defined
+    generic types, these are declared using type variables. Unlike type variables,
+    type parameters only occur in generic types.
+
+generic type
+    A generic type, or generic class, is a class with a base class
+    ``typing.Generic``, ``typing.Protocol``, or another generic type, 
+    and has at least one type parameter.
+    
+variance
+    A type parameter of a generic type can be either covariant, contravariant,
+    or invariant (the default). This defines the subtyping relation of the 
+    generic parameters and the generic type. See `PEP 484
+    <https://www.python.org/dev/peps/pep-0484/#covariance-and-contravariance>`_.
+    
+variadic
+    A variadic type parameter is either co- or contravariant. A variadic
+    type (or variadic class) is a generic type with at least one variadic 
+    type parameter
+    
+
+Rationale
+=========
+
+In PEP 484 it states:
+
+    Covariance or contravariance is not a property of a type variable,
+    but a property of a generic class defined using this variable. 
+
+More specifically, variance is a property of the *type parameters* of a 
+generic class. 
+
+Currently, variadic type parameters are declared with type varianbles
+that are constructed with ``covariant=True`` or ``contravariant=True``, 
+e.g.:
+::
+
+    from typing import TypeVar, Generic, Protocol
+    
+    T_co = TypeVar("T_co", covariant=True)
+    T_contra = TypeVar("T_contra", contravariant=True)
+    
+    class SimpleFunction(Protocol[T_contra, T_co]):
+        def __call__(self, x: T_contra) -> T_co: ...
+
+But type variables can also be used outside of the scope of generic 
+types, where the concept of variance does not apply, e.g.:
+::
+
+    def first(things: Iterable[T]) -> T:
+        return next(iter(things))
+
+
+This example is valid, but using ``T_co`` or ``T_contra`` instead of 
+``T`` here does not make sense, and type checkers do not allow is.
+
+In ``SimpleFunction``, ``T_co`` or ``T_contra`` are used to annotate
+its method as well. This is required because the variance is associated 
+with the individual type variables. But the method signature is unrelated 
+to the variance of the type parameters of its generic type, 
+``SimpleFunction`` in this example. 
+
+Because variance is restricted to the type parameters of generic types, 
+it makes more sense to specify the variance of type parameters directly
+where the generic type is defined, removing the need to specify it on 
+the type variables themselves. 
+
```

Comment:
It's strange that this isn't explicitly specified in PEP 484, but I believe that mypy does not allow this. Although you could make the case that `def f(x: T_contra) -> T_co: ...` is valid, because it is (approximately) equivalent to
```python
class F(Generic[T_contra, T_co]):
    def __call__(self, x: T_contra) -> T_co: ....

f = F()
```
But let's not overcomplicate things for now, and restrict the concept of variadic type parameters to generic types.

Reply:
PEP-484 says that variance should be ignored outside of generic classes. You reference that note earlier in your PEP.

I might be wrong, but I think `F` is different from `f` in that the type arguments of `F` are fully known before `__call__` is called, so their variance is irrelevant for `__call__`, whereas the type arguments of `f` are inferred from the call.

Reply:
I think you're right, my example is indeed different in that aspect